Minhash improvements

chris-ha458 opened the pull request:
I am beginning specific dedup method specific optimizations now.

IF/when #32 gets Pulled I will incorporate changes upon the hashing functions as well.

note : i had to upgrade mypy 1.4.1 and isort into 5.12.0.

mypy <0.980 and isort<=5.10  have bugs, especially with VS code that I cannot get around no matter how hard I try.
I am more and more encountering code that is buggy with either the old one or the new one.
So i just bit the bullet and upgraded it on my end.

the new mypy version is reporting problems with pre-commit but I'm more than happy to fix them along the way.
This might not be what you want though so let me know how you'd like to proceed.
Isort is a must but i don't mind any other mypy version.

chris-ha458 commented:
`Hs = [bytes(hashvalues[start:end].byteswap().data) for start, end in hashranges]`
Why is byteswap necessary here? 

I guessed that you are "recovering" original big-endian byte order since sha1 hashes are read little-endian.
however, the original datasketch does not use it, and when I remove it it doesnt change downstream behavior.
Was it necessary from testing on your end?
Anyway, if that was indeed the reason, I will remove it from any implementation that uses xxh since they output big-endian anyway.

chris-ha458 commented:
Also, this is representative of what I hope to do for simhash, ccnet etc
(of course this is addition to potential hashing improvements)

ChenghaoMou commented:
To answer you question, it is actually used in `datasketch` (see [search results](https://github.com/search?q=repo:ekzhu/datasketch%20byteswap&type=code)):
![CleanShot 2023-08-03 at 18 36 12@2x](https://github.com/ChenghaoMou/text-dedup/assets/32014649/abb90508-1f94-4cdd-992d-aaa434671985)

I think it is actually agnostic to the endianness at that stage. I just kept it intact copying from datasketch.

ChenghaoMou reviewed (approved):
LGTM. Thanks for making it better!

chris-ha458 commented:
it seems like you approved it but due to the hashing improvements being pulled this needs resolution.
I will work on that.

as for byteswap : `https://github.com/ekzhu/datasketch/issues/114`

1. i will investigate if having that is indeed faster (due to some internal representation? i can't see why that would be true for numpy based codebase currently)
2. if it is not faster, check if removing it harms backward compatibility (again at that point I cannot see why it would. we would just be comparing them not using them for any calc)
3. If it is faster, check if it is possible to utilize bitarrays which would be faster and lower footprint for such operations
4. If it is faster, find a way to take it outside the hot loop (possible byteswap all at once instead of in the middle of the hotloop)

chris-ha458 commented:
merging done!
i'll work on more of it in the meantime.

chris-ha458 commented:
> * i will investigate if having that is indeed faster (due to some internal representation? i can't see why that would be true for numpy based codebase currently)
> * if it is not faster, check if removing it harms backward compatibility (again at that point I cannot see why it would. we would just be comparing them not using them for any calc)
> * If it is faster, check if it is possible to utilize bitarrays which would be faster and lower footprint for such operations
> * If it is faster, find a way to take it outside the hot loop (possible byteswap all at once instead of in the middle of the hotloop)

* Speed. it is neutral. the total difference due to using or not using byteswap is not detectable with statistical significance.
This is astonishing since it is 1. in the hot loop(embed func) 2. another hot loop (within a string related for loop list comprehension). I guess the function itself is highly optimized that it is negligible.
* backward compatibility. for the datasets I've tested (oscar 2201, 2301, multiple languages) it yield equal results

I won't touch it for now. even if there is any gains to be made it would be at best marginal.

chris-ha458 commented:
I went on and implemented lower precision hashes as well.
using xxh3, 16bit hashes , minhashing processing time and memory consumption drops to around 50%
(twice as faster)
However, this does cause a bit of more false positives (but also less false negatives) and I think we can adjust permutations, bands and rows to recover this. its usually around 1% difference

```
OSCAR-2301 "af" 64bits (default)
MinHashing                      : 8.40s
Before                          : 11997
After                           : 8637
32 bits. (theoretically minimal differences with 50% memory reduction since even for default 64bits we only use 32bit hashes)
MinHashing                      : 5.23s (38% reduction)
After                           : 8615 (0.2% difference)
16 bits
MinHashing                      : 3.73s (56% reduction)
After                           : 8490 (1.7%)
```
I feel like this will enable much larger datasets with more tradeoff options to the enduser.

ChenghaoMou commented on text_dedup/minhash.py:
```diff
@@ -32,8 +31,10 @@
+from text_dedup.utils.hashfunc import xxh3_32hash
 from text_dedup.utils.timer import Timer
 
 SEED = 42
```
Maybe remove the `MAX_HASH` and the `MERSENNE_PRIME` here?

ChenghaoMou commented on text_dedup/minhash.py:
```diff
@@ -99,94 +100,39 @@ def embed_func(
     ... ).T
-    >>> res = embed_func(content, idx, num_perm=num_perm, ngram_size=ngram_size, min_length=0, hashranges=hashranges, 
+    >>> res = embed_func(content, idx, num_perm=num_perm, ngram_size=ngram_size, min_length=0, hashranges=hashranges,
     ... permutations=PERMUTATIONS, hash_func=xxh3_hash)
```
`xxh3_hash` becomes invalid since the import change.

ChenghaoMou replied:
For local testing purposes, you can run `pytest --doctest-modules . --ignore deduplicate-text-datasets` for a quick sanity check.

chris-ha458 replied:
thanks for the tip! I have been using run testing (running the scripts to actually process the data itself)
but this would be helpful for quick tests!

ChenghaoMou reviewed (commented):
Thank you! Only two minor issues.

chris-ha458 commented:
i fixed the docstring but that turned out to reveal a deeper issue with the function.
to avoid scope conflicts I added them into the function definitions.

I could pass in a dict/list/tuple but the way i did it is faster (i think).
if you feel the need to change it, i think i can make a tuple called HASH_CONFIG and pass it, and slice it.

let me know what you think

ChenghaoMou reviewed (approved):
LGTM. I also tested the change with the benchmark in `benchmarks/pinecone.ipynb`:
```
no change (sha1 + 64bit):
{
'Correct': 92401,
'Incorrect': 7599,
'Accuracy': 0.924,
'Recall': 0.9646,
'Precision': 0.4434
}
 
xxh3 + 16 bit
{
'Correct': 91746,
'Incorrect': 8254,
'Accuracy': 0.9175,
'Recall': 0.9662,
'Precision': 0.4419
}

xxh3 + 32 bit
{
'Correct': 92630,
'Incorrect': 7370,
'Accuracy': 0.9263,
'Recall': 0.9667,
'Precision': 0.446
}

xxh3 + 64 bit
{
'Correct': 92526,
'Incorrect': 7474,
'Accuracy': 0.9253,
'Recall': 0.9671,
'Precision': 0.4443
 }
```
xxh3 does offer some more options in terms of the tradeoff between speed and performance. (Time is omitted here because it was only a single run, but xxh3 16-bit was indeed faster in the test)

chris-ha458 commented:
Thanks! Ill use the pinecone for future pr and testing as well.

Now that it is approved, is there anything I need to do on my end for it to merge?
